---
title: "Day 2 - Module 08: Count normalizations and DESeq2 model fitting"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            theme: readable
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            keep_md: true
            markdown: GFM
---

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>
> # Objectives
> * Discuss count normalizations
> * Execute model fitting for differential expression comparisons

```{r LoadRunningData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
load("RunningData.RData")
```


**TODO: Add schedule to top**

# Count normalizations

Since counts of mapped reads for each gene is proportional to the expression of RNA in addition to many “uninteresting” other factors, normalization is the process of scaling raw count values to account for the “uninteresting” factors and ensure expression levels are more comparable.

## Normalization goals

Two common factors that need to be accounted for during normalization are **sequencing depth** and **gene length**.    

* **Sequencing depth** normalization is neccessary to account for the proportion of reads per gene expected for more deeply sequenced samples (like in pink below) versus a less deeply sequenced sample (like in green blow.)    

![*Note that each pink or green rectangle represents an aligned read, with reads spanning an intron connected by a dashed line.Figure credit to [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](~/repos/rnaseq_demystified_workshop_2021/images/normalization_methods_depth.png)    

    
* **Gene length** normalization is necessary since genes of different lengths have different probablities of generating fragments that end up in the library. In the example below, both genes have similar levels of expression. However, the number of reads that map to the longer gene (Gene X) will be much great than the number of reads that map to the short gene (Gene Y).    

![*Each pink rectangle represents an aligned read, with reads spanning an intron connected by a dashed line.Figure credit to [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](~/repos/rnaseq_demystified_workshop_2021/images/normalization_methods_length.png)      
    
> **Note**: The above figures are originally from a [HBC  tutorial](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html) that also includes a detailed comparison of different normalization (CPM, TPM, FPKM) approaches and their best uses.

## DESeq2 normalizations

An additional consideration for normalization is **RNA composition**. A few highly differentially expressed genes, differences in the number of genes expressed between samples, or contamination are not accounted for by depth or gene length normalization methods. Accounting for RNA composition is particularly important for differential expression analyses, regardless of the tool used.

![*Figure from [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](~/repos/rnaseq_demystified_workshop_2021/images/normalization_methods_composition.png)

    
DESeq2 has an [internal normalization process](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106). However for data exploration and visualizations, it is helpful to generate an object of independently normalized counts.

For downstream quality control visualization, we will use the [rlog transformation](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations).


```{r CountNorm}
rld <- rlog(dds, blind = TRUE)
```

The rlog transformation produces log2 scale data that has also been normalized to overall library size as well as variance across genes at different mean expression levels. For larger numbers of samples, there is an alternative transformation method, [vst](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations) that can be used instead for count normalizations. 


```{r CountNormCheck}
head(assay(rld), 3)
```

Looking at the rld values, we can see that they are now in log scale. Since we set `blind=TRUE`, the transformation is blind to the sample information we specified in the design formula. The normalized counts are only necessary for visualization methods during expresion-level quality assessment and aren't used in the model fitting. 

# DESeq2 Model Fitting

Next, we'll fit our standard model and our model that includes the patient origin covariate using the `DESeq` function and take a look at the objects we generate. 
```{r FitModelStandard, message=FALSE}
# Apply model - takes some time to run
dds <- DESeq(dds)
```
```{r FitModelStandardCheck1, eval=TRUE}
resultsNames(dds)
```

The results include three pairwise comparison to specified control as default but other information is now stored in the `dds` object so can generate additional pairwise comparisons.

```{r FitModelStandardCheck2, eval=FALSE}
head(dds)
```

Optional - we can fit a seperate DESeq2 model for our patient sample example, which included a covariate in our model. 
```{r FitModelCovariate, message=FALSE}
dds_patient <- DESeq(dds_patient)
```
```{r FitModelCovariateCheck1,eval=TRUE}
resultsNames(dds_patient)
```

Notice that with the additional covariate, additional comparisons are generated by the tool. Since we arbitrarily added the patient origin information, the `dds` object will be what we primarily use moving forward. You can explore the impact of adding a covariate by substituting `dds_patient` for `dds` when working through later excercises on your own since both DESeq2 objects have their data organized in the same way.

**Checkpoint**: *If you see the same results when you execute `resultsNames(dds)`, please indicate with the green 'yes' button. Otherwise, please use the 'red x' button to get help*


**Before moving on with our group comparisons, we'll take a short (10 minute) break.**

---

# Sources
## Training resources used to develop materials
* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html



```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "RunningData.RData")
```


These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
